service/agentconf.go

```diff
@@ -0,0 +1,274 @@
+}
+
+// WriteSystemdAgents creates systemd files and create symlinks for the list of machine and units passed in the standard filepath '/var/lib/juju' during the upgrade process.
+func WriteSystemdAgents(machineAgent string, unitAgents []string, dataDir string, symLinkSystemdDir string, symLinkSystemdMultiUserDir string, series string) (errSysServiceNames []string, errSymServiceNames []string, errAgentNames []string, lastError error) {
+
+	for _, agentName := range append(unitAgents, machineAgent) {
+		conf, err := CreateAgentConf(agentName, dataDir, series)
+		if err != nil {
+			logger.Infof("%s", err)
+			lastError = err
+			continue
+		}
+
+		svcName := "jujud-" + agentName
+		svc, err := NewService(svcName, conf, series)
+		if err != nil {
+			logger.Infof("Failed to create new service %s: ", err)
+			continue
+		}
+
+		upSvc, ok := svc.(UpgradableService)
+		if !ok {
+			initName, err := VersionInitSystem(series)
+			if err != nil {
+				return nil, nil, nil, errors.Trace(errors.Annotate(err, "nor is service an UpgradableService"))
+			}
+			return nil, nil, nil, errors.Errorf("%s service not of type UpgradableService", initName)
+		}
+
+		if err = upSvc.WriteService(); err != nil {
+			logger.Infof("failed to write service for %s: %s", agentName, err)
+			errAgentNames = append(errAgentNames, agentName)
+			lastError = err
+			continue
+		}
+
+		running, err := systemd.IsRunning()
+		switch {
+		case err != nil:
+			return nil, nil, nil, errors.Errorf("failure attempting to determine if systemd is running: %#v\n", err)
+		case running:
+			// Links for manual and automatic use of the service
+			// have been written, move to the next.
+			errSysServiceNames = append(errSysServiceNames, svcName)
+			logger.Infof("wrote %s agent, enabled and linked by systemd", svcName)
+			continue
+		}
+
+		svcFileName := svcName + ".service"
+		if err = os.Symlink(path.Join(dataDir, "init", svcName, svcFileName),
+			path.Join(symLinkSystemdDir, svcFileName)); err != nil && !os.IsExist(err) {
+			return nil, nil, nil, errors.Errorf("failed to link service file (%s) in systemd dir: %s\n", svcFileName, err)
+		}
+
+		if err = os.Symlink(path.Join(dataDir, "init", svcName, svcFileName),
+			path.Join(symLinkSystemdMultiUserDir, svcFileName)); err != nil && !os.IsExist(err) {
+			return nil, nil, nil, errors.Errorf("failed to link service file (%s) in multi-user.target.wants dir: %s\n", svcFileName, err)
+		}
+
+		errSymServiceNames = append(errSymServiceNames, svcName)
+		logger.Infof("wrote %s agent, enabled and linked by symlink", svcName)
+	}
+	return errSysServiceNames, errSymServiceNames, errAgentNames, lastError
+}
+
+// CreateAgentConf creates the configfile for specified agent running on a host with specified series.
+func CreateAgentConf(agentName string, dataDir string, series string) (_ common.Conf, err error) {
+	defer func() {
+		if err != nil {
+			errors.Errorf("failed create agent conf for %s: %s", agentName, err)
```

Comment:
This needs to be 
```go
err = errors.Errorf("failed create agent conf for %s: %s", agentName, err)
```

Otherwise the new error value will be created and then just thrown away - the original one would still be returned. It might be better to use errors.Annotatef here, too.